openedx/core/djangoapps/oauth_dispatch/models.py

```diff
@@ -43,3 +152,82 @@ def verify_access_token_as_expired(cls, access_token):
         is set at the beginning of the epoch which is Jan. 1, 1970
         """
         return access_token.expires == datetime(1970, 1, 1, tzinfo=utc)
+
+
+class OauthRestrictedApplication(AbstractApplication):
+    """
+    Application model for use with Django OAuth Toolkit that allows the scopes
+    available to an application to be restricted on a per-application basis.
+    """
+    allowed_scope = models.TextField(blank = True)
+
+    def _get_list_from_delimited_string(self, delimited_string, separator=_DEFAULT_SEPARATOR):
+        """
+        Helper to return a list from a delimited string
+        """
+
+        return delimited_string.split(separator) if delimited_string else []
+
+    @classmethod
+    def is_token_oauth_restricted_application(cls, token):
+        """
+        Returns if token is issued to a RestriectedApplication
+        """
+
+        if isinstance(token, basestring):
+            # if string is passed in, do the look up
+            token_obj = AccessToken.objects.get(token=token)
+        else:
+            token_obj = token
+
+        return cls.get_restricted_application(token_obj.application) is not None
+
+    @classmethod
+    def get_restricted_application(cls, application):
+        """
+        For a given application, get the related restricted application
+        """
+        return OauthRestrictedApplication.objects.filter(id=application.id)
+
+    @property
+    def allowed_scopes(self):
+        """
+        Translate space delimited string to a list
+        """
+        all_scopes = set(get_scopes_backend().get_all_scopes().keys())
+        app_scopes = set(self._get_list_from_delimited_string(self.allowed_scope))
+        return app_scopes.intersection(all_scopes)
+
+    @classmethod
+    def set_access_token_as_expired(cls, instance):
+        """
+        For access_tokens for RestrictedApplications, put the expire timestamp into the beginning of the epoch
+        which is Jan. 1, 1970
+        """
+        
+        instance.expires = datetime(1970, 1, 1, tzinfo=utc)
```

Comment:
Can you explain what this new model is meant to be responsible for?  Don't we already have the `RestrictedApplication` class for the expired tokens?

Reply:
Based on previous discussion, I thought we are going to remove RestrictedApplication. Hence added expired tokens method in the new model.  Now, its reverted back.